src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs

```diff
@@ -0,0 +1,479 @@
+        /// </exception>
+        /// <remarks>
+        /// This encodes an attribute with the OID 0.9.2342.19200300.100.1.25 as an IA5String.
+        /// </remarks>
+        public void AddDomainComponent(string domainComponent)
+        {
+            // RFC 5912
+            // id-domainComponent      AttributeType ::=
+            //  { itu-t(0) data(9) pss(2342) ucl(19200300) pilot(100)
+            //  pilotAttributeType(1) 25 }
+            //  at-domainComponent ATTRIBUTE ::= {TYPE IA5String
+            //  IDENTIFIED BY id-domainComponent }
+
+            ArgumentException.ThrowIfNullOrEmpty(domainComponent);
+            EncodeComponent(Oids.DomainComponent, domainComponent, UniversalTagNumber.IA5String);
+        }
+
+        /// <summary>
+        /// Builds an <see cref="X500DistinguishedName" /> that represents the encoded attributes.
+        /// </summary>
+        /// <returns>
+        /// An <see cref="X500DistinguishedName" /> that represents the encoded attributes.
+        /// </returns>
+        public X500DistinguishedName Build()
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+
+            using (writer.PushSequence())
+            {
+                foreach (byte[] component in _encodedComponents)
+                {
+                    writer.WriteEncodedValue(component);
+                }
+            }
+
+            byte[] rented = CryptoPool.Rent(writer.GetEncodedLength());
+            int encoded = writer.Encode(rented);
+            X500DistinguishedName name = new X500DistinguishedName(rented.AsSpan(0, encoded));
+            CryptoPool.Return(rented, clearSize: 0); // Distinguished Names do not contain sensitive information.
+            return name;
+        }
+
+        private void EncodeComponent(
+            string oid,
+            string value,
+            UniversalTagNumber stringEncodingType,
+            [CallerArgumentExpression("value")] string? paramName = null)
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+
+            using (writer.PushSetOf())
+            using (writer.PushSequence())
+            {
+                writer.WriteObjectIdentifier(oid);
+
+                try
+                {
+                    writer.WriteCharacterString(stringEncodingType, value);
+                }
+                catch (EncoderFallbackException)
+                {
+                    throw new ArgumentException(SR.Format(SR.Argument_Asn1_InvalidStringContents, stringEncodingType), paramName);
+                }
+            }
+
+            _encodedComponents.Add(writer.Encode());
+        }
+
+        private void EncodeComponent(
+            string oid,
+            ReadOnlySpan<byte> value,
+            [CallerArgumentExpression("value")] string? valueParamName = null)
+        {
+            if (!AsnDecoder.TryReadEncodedValue(value, AsnEncodingRules.DER, out _, out _, out _, out _))
```

Comment:
If skipping the length check is intentional (so that "invalid DER" and "one value at a time" are different messages) then that's probably worth a comment.

Otherwise, it feels like we should check that bytesConsumed == value.Length or throw the message controlled at this layer.

Reply:
I reversed course here after adding a few tests.